lib/transport/JsonRpcTransport/JsonRpcTransport.js

```diff
@@ -0,0 +1,104 @@
+const MaxRetriesReachedError = require('../errors/MaxRetriesReachedError');
+const NoAvailableAddressesForRetry = require('../errors/NoAvailableAddressesForRetry');
+
+class JsonRpcTransport {
+  /**
+   * @param {createAddressProviderFromOptions} createAddressProviderFromOptions
+   * @param {requestJsonRpc} requestJsonRpc
+   * @param {ListAddressProvider|SMLAddressProvider|AddressProvider} addressProvider
+   * @param {DAPIClientOptions} globalOptions
+   */
+  constructor(
+    createAddressProviderFromOptions,
+    requestJsonRpc,
+    addressProvider,
+    globalOptions,
+  ) {
+    this.createAddressProviderFromOptions = createAddressProviderFromOptions;
+    this.requestJsonRpc = requestJsonRpc;
+    this.addressProvider = addressProvider;
+    this.globalOptions = globalOptions;
+
+    this.lastUsedAddress = null;
+  }
+
+  /**
+   * Make request to DAPI node
+   *
+   * @param {string} method
+   * @param {object} [params]
+   * @param {DAPIClientOptions} [options]
+   *
+   * @returns {Promise<object>}
+   */
+  async request(method, params = {}, options = { }) {
+    const addressProvider = this.createAddressProviderFromOptions(options)
+      || this.addressProvider;
+
+    const address = await addressProvider.getLiveAddress();
+
+    this.lastUsedAddress = address;
+
+    // eslint-disable-next-line no-param-reassign
+    options = {
+      retries: this.globalOptions.retries,
+      timeout: this.globalOptions.timeout,
+      ...options,
+    };
+
+    const requestOptions = {};
```

Comment:
I think you can just pass timeout as previously coz you are checking for undefined already in `requestJsonRpc`

Reply:
maybe it's better to keep them independent?

Reply:
Yeah, what if `timeout` is `null`? Maybe we should introduce `validateRequestOptions` method and use it here and in the `DAPIClient#constructor`?

Reply:
Just forget about that :)